File: javascript/dataStructures/immutableLinkedListFunc.js
// !!! head ant tail need to thrw an exception then called on an empty list! and your code needs to 
// change to reflect this change
// !!! implement tail recursive alternatives when possible

/**
*
* Constructing Linked Lists
*
* Our linked list data structure consists of two fundamental building blocks: Nil and cons. Nil represents the 
* empty list and serves as a sentinel for more complex lists. The cons operation extends a list at the front by 
* inserting a new value.
*
* The lists we construct using this method consist internally of nested arrays of 2 items. For example, the list [1, 2, 3] 
* is represented by the expression cons(1, cons(2, cons(3, Nil))) which evaluates to the nested arrays [1, [2, [3, []]]].
*
*/

var Nil = [];

/** Extends a list at the front by inserting a new value. */
function cons(head, tail) {
    tail = typeof tail === "undefined" ? Nil : tail;  // default argument is Nil
    return [head, tail];
}

/**
* Define list instances using a more convenient syntax and without deeply nested cons calls.
* lst() === Nil                                                                                                                                                                
* lst(1, 2, 3) === [1, [2, [3, []]]]
*/
function lst() {
    var arg_array = Array.prototype.slice.call(arguments);  // arguments cast from object to array
    if (arg_array.length === 0) return Nil;
    else return cons(arg_array[0], lst.apply(this, arg_array.slice(1)));
}

/** Returns the first element of a list */
function head(xs) {
    if (xs === Nil) return [];
    else return xs[0];
}

/** Returns a list containing all elements except the first. */
function tail(xs) {
    if (xs === Nil) return [];
    else return xs[1];
}

/** Returns True if the list contains zero elements. */
function isEmpty(xs) {
    return xs === Nil;
}

/** Returns number of elements in a given list. */
function length(xs) {
    if (isEmpty(xs)) return 0;
    else return 1 + length(tail(xs));
}

/** Concatenates two lists. */
function concat(xs, ys) {
    if (isEmpty(xs)) return ys;
    else return cons(head(xs), concat(tail(xs), ys));
}

/** Returns the last element of a non-empty list. */
function last(xs) {
    if (isEmpty(tail(xs))) return head(xs);
    else return last(tail(xs));
}

/** Returns all elements except the last one. */
function init(xs) {
    if (isEmpty(tail(tail(xs)))) return cons(head(xs));
    else return cons(head(xs), init(tail(xs)));
}

/** Returns the input list, reversed. */
function reverse(xs) {
    if (isEmpty(xs)) return xs;
    else return concat(reverse(tail(xs)), cons(head(xs), Nil));
}

/** Returns the first n elements of the given list. Inclusive, one-based indexing. */
function take(n, xs){
    if (n < 1) return Nil;
    else return cons(head(xs), take(n - 1, tail(xs)));
}

/** Returns a copy of input list, without the first n elements. Inclusive, one-based indexing. */
function drop(n, xs) {
    if (n < 1) return xs;
    else return drop(n - 1, tail(xs));
}

/** 
* Returns a subset of the input list that includes the items in the range from, to. 
* Zero-based indexing, from is inclusive, to is exclusive.  
*/
function piece(from, to, xs) {
    return take((to - from), drop(from, xs));
}

/** Removes the element of the list at given index. One-based indexing. */
function removeOne(index, xs) {
    return concat(take(index - 1, xs), drop(index, xs));
}

/** Returns the value at n index. Zero-based indexing. */
function at(index, xs){
    if (index < 1) return head(xs);
    else return(at(index - 1, tail(xs)));
}

// Some examples:

var a = lst(1, 2, 3, 4, 5);
var b = cons(6, cons(7, cons(8, cons(9, cons(10)))));
var c = lst(12, 3, 16, 22, 89, 4, 1);
var empty = lst();

console.log("a:               ", JSON.stringify(a));
console.log("b:               ", JSON.stringify(b));
console.log("empty:           ", JSON.stringify(empty));
console.log("head(a):         ", JSON.stringify(head(a)));
console.log("tail(a):         ", JSON.stringify(tail(a)));
console.log("tail(empty):     ", JSON.stringify(tail(empty)));
console.log("concat(a, b):    ", JSON.stringify(concat(a, b)));
console.log("take(3, a):      ", JSON.stringify(take(3, a)));
console.log("drop(3, a):      ", JSON.stringify(drop(3, a)));
console.log("piece(0, 4):     ", JSON.stringify(piece(0, 4, a)));
console.log("removeOne(3, a): ", JSON.stringify(removeOne(3, a)));
console.log("at(2, a):        ", JSON.stringify(at(2, a)));

// !!! still needs to_string and isort functions

File: python/DataStructures/Lists/immutable_linked_list_func.py
# http://dbader.org/blog/functional-linked-lists-in-python
"""
Constructing linked lists

Our linked list data structure consists of two fundamental building blocks: Nil and cons. Nil represents the 
empty list and serves as a sentinel for more complex lists. The cons operation extends a list at the front by 
inserting a new value.

The lists we construct using this method consist internally of nested 2-tuples. For example, the list [1, 2, 3] 
is represented by the expression cons(1, cons(2, cons(3, Nil))) which evaluates to the nested tuples (1, (2, (3, Nil))).
"""

Nil = None

def cons(head, tail=Nil):
    return (head, tail)

"""
Abstracting away the tuple construction using the cons function gives us a lot of flexibility on how lists are 
represented internally as Python objects. For example, instead of using 2-tuples we could store our elements in 
a chain of anonymous functions using Pythons lambda keyword.
"""
def lst(*xs):
    """
    Define list instances using a more convenient syntax and without deeply nested cons calls.
    lst() == Nil
    lst(1, 2, 3) == (1, (2, (3, Nil))) 
    """
    if not xs:
        return Nil
    else:
        return cons(xs[0], lst(*xs[1:]))

def head(xs):
    """ Returns the frst element of a list. """
    return xs[0]

def tail(xs):
    """ Returns a list containing all elements except the first. """
    return xs[1]

def is_empty(xs):
    """ Returns True if the list contains zero elements. """
    return xs is Nil

def length(xs):
    """
    Returns number of elements in a given list. To find the length of a list we need to scan all of its
    elements, thus leading to a time complexity of O(n).
    """
    if is_empty(xs):
        return 0
    else:
        return 1 + length(tail(xs))

def concat(xs, ys):
    """ Concatenates two lists. O(n) """
    if is_empty(xs):
        return ys
    else:
        return cons(head(xs), concat(tail(xs), ys))

def last(xs):
    """ Returns last element of a non-empty list. O(n) """
    if is_empty(tail(xs)):
        return head(xs)
    else:
        return last(tail(xs))

def init(xs):
    """ Returns all elements except the last one. O(n) """
    if is_empty(tail(tail(xs))):
        return cons(head(xs))
    else:
        return cons(head(xs), init(tail(xs)))

def reverse(xs):
    """ Returns the input list, reversed. O(n*2) """
    if is_empty(xs):
        return xs
    else:
        return concat(reverse(tail(xs)), cons(head(xs), Nil))

def take(n, xs):
    """ Returns the first n elements of the list. """
    if n < 1:
        return Nil
    else:
        return cons(head(xs), take(n - 1, tail(xs)))

def drop(n, xs):
    """ Returns a copy of input list, without the first n elements."""
    if n < 1:
        return xs
    else:
        return drop( n - 1, tail(xs))

def piece(from_index, to_index, xs):
    """ Returns a subset of the input list that includes the items in the range from_index, to_index """ 
    return take(to_index - from_index, drop(from_index, xs))

def remove_one(index, xs):
    """ Returns a copy of the input list which doesn't contain the node in the specified index """
    return concat(take(index - 1, xs), drop(index, xs))

def to_string(xs, prefix="[", sep=", ", postfix="]"):
    """ lst to string """
    def _to_string(xs):
        if is_empty(xs):
            return ""
        elif is_empty(tail(xs)):
            return str(head(xs))
        else:
            return str(head(xs)) + sep + _to_string(tail(xs))
    return prefix + _to_string(xs) + postfix

def insert(x, xs):
    """ Helper for functional insertion sort. """
    if is_empty(xs) or x <= head(xs):
        return cons(x, xs)
    else:
        return cons(head(xs), insert(x, tail(xs)))

def isort(xs):
    """ Functional insertion sort """
    if is_empty(xs):
        return xs
    else:
        return insert(head(xs), isort(tail(xs)))

def at(index, xs):
    """ Returns the value at index n. Zero-based indexing. """
    if index < 1:
        return head(xs)
    else:
        return at(index - 1, tail(xs))

# Examples:

y = cons(Nil)
z = lst()
w = lst(Nil)
a = lst(1, 2, 3, 4)
b = lst(5, 6, 7, 8)
ab = concat(a, b)
print y
print z
print w
print ""
print a
print head(a)
print tail(a)
print is_empty(a)
print length(a)
print last(ab)
print init(ab)
print reverse(ab)

print remove_one(1, a)
print take(4, ab)
print remove_one(4, ab)
print ""
print drop(3, ab)
print ""
print piece(0, 5, ab)
print piece(1, 5, ab)
print piece(2, 5, ab)
print piece(1, 6, ab)
print piece(1, 7, ab)
print at(2, a)

